src/backend/distributed/transaction/backend_data.c

```diff
@@ -695,9 +699,22 @@ InitializeBackendData(void)
 
 	LockBackendSharedMemory(LW_EXCLUSIVE);
 
-	/* zero out the backend data */
+	/* zero out the backend its transaction id */
 	UnSetDistributedTransactionId();
-	UnSetGlobalPID();
+
+	/*
+	 * Use the given globalPID to initialize
+	 */
+	if (globalPID == INVALID_CITUS_INTERNAL_BACKEND_GPID)
+	{
+		MyBackendData->distributedCommandOriginator =
```

Comment:
why not always set `MyBackendData->globalPID`

MyBackendData->globalPID = globalPID
MyBackendData->distributedCommandOriginator = (globalPID == INVALID_CITUS_INTERNAL_BACKEND_GPID) ? false : true;

or smth

Reply:
I mainly wanted to keep the old behaviour of having 0 as the globalPID. But this made me realize that I removed the UnSetGlobalPID call. I added that one back.

Furthermore I made two small changes:
1. I also added some locks around changing the distributedCommandOriginator/globalPid
2. I swapped the `SetActiveMyBackend` and `UnsetGlobalPID` calls, so that a backend is marked as inactive before unsetting the global PID (this might have been another cause of the race).

Reply:
> I also added some locks around changing the distributedCommandOriginator/globalPid

Ops, good point

> I swapped the SetActiveMyBackend and UnsetGlobalPID calls, so that a backend is marked as inactive before unsetting the global PID (this might have been another cause of the race).

We have way too many functions for doing similar things. I'll try to open another PR to clean up unnecessary logic, but for now, this makes sense